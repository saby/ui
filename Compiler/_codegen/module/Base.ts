/**
 */

import {
    IDependenciesController,
    createDependenciesController,
} from './Dependencies';
import { IExportsController, createExportsController } from './Exports';

/**
 * Интерфейс процессора, который выполняет компиляцию модуля в различный формат.
 */
export interface IModuleProcessor {
    /**
     * Установить имя компилируемого модуля.
     * @param name {string} Имя компилируемого модуля, которое содержит имя плагина.
     */
    setModuleName(name: string): void;

    /**
     * Установить флаг строго режима для компилируемого модуля.
     * @param value {boolean} Значение флага строго режима.
     */
    setStrictMode(value: boolean): void;

    /**
     * Добавить зависимость модуля. Если не указано имя зависимости, то зависимость считается неименованной.
     * @param ref {string} Путь зависимости.
     * @param name {string?} Имя зависимости.
     */
    addDependency(ref: string, name?: string): void;

    /**
     * Добавить блок код к телу компилируемого модуля.
     * @param block {string} Блок кода, который составляет тело модуля.
     */
    addCodeBlock(block: string): void;

    /**
     * Установить блок кода, который будет возвращен из модуля как экспортируемые данные.
     * @param block {string} Блок кода как экспортируемые данные.
     */
    setReturnableExport(block: string): void;

    /**
     * Выполнить компиляцию модуля.
     * @returns {string} JavaScript код модуля.
     */
    compile(): string;
}

/**
 * Абстрактный базовый класс процессора модулей.
 * Предоставляет общие данные и методы для выполнения компиляции.
 */
export abstract class Base implements IModuleProcessor {
    /**
     * Имя компилируемого модуля.
     * @protected
     */
    protected name: string;

    /**
     * Флаг включения строго режима для компилируемого модуля.
     * @protected
     */
    protected useStrict: boolean;

    /**
     * Массив кода, который составляет тело будущего модуля.
     * @protected
     */
    protected code: string[];

    /**
     * Инстанс контроллера зависимостей, необходимый для работы с именованными и
     * безымянными (анонимными) зависимостями модуля.
     * @protected
     */
    protected dependenciesController: IDependenciesController;

    /**
     * Инстанс контроллера экспортируемых данных.
     * @protected
     */
    protected exportsController: IExportsController;

    /**
     * Инициализировать инстанс базового класса.
     * @protected
     */
    protected constructor() {
        this.name = '';
        this.useStrict = false;
        this.code = [];
        this.dependenciesController = createDependenciesController();
        this.exportsController = createExportsController();
    }

    /**
     * Установить имя компилируемого модуля.
     * @param name {string} Имя компилируемого модуля, которое содержит имя плагина.
     */
    setModuleName(name: string): void {
        this.name = name;
    }

    /**
     * Установить флаг строго режима для компилируемого модуля.
     * @param value {boolean} Значение флага строго режима.
     */
    setStrictMode(value: boolean): void {
        this.useStrict = value;
    }

    /**
     * Добавить зависимость модуля. Если не указано имя зависимости, то зависимость считается неименованной.
     * @param ref {string} Путь зависимости.
     * @param name {string?} Имя зависимости.
     */
    addDependency(ref: string, name?: string): void {
        this.dependenciesController.addDependency(ref, name);
    }

    /**
     * Добавить блок код к телу компилируемого модуля.
     * @param block {string} Блок кода, который составляет тело модуля.
     */
    addCodeBlock(block: string): void {
        this.code.push(block);
    }

    /**
     * Установить блок кода, который будет возвращен из модуля как экспортируемые данные.
     * @param block {string} Блок кода как экспортируемые данные.
     */
    setReturnableExport(block: string): void {
        this.exportsController.setReturnableExport(block);
    }

    /**
     * Выполнить компиляцию модуля.
     * @abstract
     * @returns {string} JavaScript код модуля.
     */
    abstract compile(): string;
}
