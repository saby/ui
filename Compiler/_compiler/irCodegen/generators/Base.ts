/**
 * @author Krylov M.A.
 *
 * Базовый генератор обращений к именованным сущностям.
 */

import type { TPrimitive, IFormatter } from '../Interface';
import type { TAliasesMap } from '../Aliases';

import { wrapSequence } from '../types/String';

export default class Base {
    protected constructor(
        protected readonly formatter: IFormatter,
        protected readonly instance: string,
        protected readonly aliases: TAliasesMap,
        protected readonly isRelease: boolean
    ) { }

    protected getPropertyName(fn: string): string {
        if (this.isRelease) {
            if (this.aliases.has(fn)) {
                return `.${this.aliases.get(fn)}/*${fn}*/`;
            }
        }

        return `.${fn}`;
    }

    protected generateMethodCall(method: string, args: TPrimitive[]): string {
        return `${this.instance}${this.getPropertyName(method)}(${wrapSequence(args)})`;
    }

    protected generateFormattedMethodCall(method: string, args: TPrimitive[]): string {
        return `${this.instance}${this.getPropertyName(method)}(${this.formatter.formatSequence(args, 1)})`;
    }
}
