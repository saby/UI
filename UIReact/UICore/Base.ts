export { default as Control}  from './_base/Control';

export {IControlOptions, TemplateFunction} from './_base/interfaces';
