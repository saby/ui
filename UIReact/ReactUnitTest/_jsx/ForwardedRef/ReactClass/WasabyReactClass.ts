import { Control, TemplateFunction } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_jsx/ForwardedRef/ReactClass/WasabyReactClass';

export default class WasabyReactClass extends Control {
    protected _template: TemplateFunction = template;
    protected _hasContainer: boolean = false;

    protected _afterMount() {
        this._hasContainer = !!this._container;
    }
}
