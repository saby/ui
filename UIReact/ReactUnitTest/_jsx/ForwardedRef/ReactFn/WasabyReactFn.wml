<ws:if data="{{ _options.contentOption }}">
    <ReactUnitTest._jsx.ForwardedRef.ReactFn.ReactFnContent>
        <ReactUnitTest._jsx.ForwardedRef.ReactFn.ReactFn hasContainer="{{ _hasContainer }}"/>
    </ReactUnitTest._jsx.ForwardedRef.ReactFn.ReactFnContent>
</ws:if>
<ws:else>
    <ReactUnitTest._jsx.ForwardedRef.ReactFn.ReactFn hasContainer="{{ _hasContainer }}" />
</ws:else>
