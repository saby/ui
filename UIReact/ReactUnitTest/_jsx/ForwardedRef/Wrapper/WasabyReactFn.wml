<ReactUnitTest._jsx.ForwardedRef.Wrapper.ReactFn
    content="{{ _options.content}}"
    hasContainer="{{ _options.hasContainer }}" />
