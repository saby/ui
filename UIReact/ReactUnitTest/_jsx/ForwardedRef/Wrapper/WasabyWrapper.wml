<ws:if data="{{ _options.contentOption }}">
    <ReactUnitTest._jsx.ForwardedRef.Wrapper.WasabyReactFn>
        <ReactUnitTest._jsx.ForwardedRef.Wrapper.WasabyReactFn>
            <div>{{ _hasContainer ? 'ok' : 'not ok' }}</div>
        </ReactUnitTest._jsx.ForwardedRef.Wrapper.WasabyReactFn>
    </ReactUnitTest._jsx.ForwardedRef.Wrapper.WasabyReactFn>
</ws:if>
<ws:else>
    <ReactUnitTest._jsx.ForwardedRef.Wrapper.WasabyReactFn>
        <ReactUnitTest._jsx.ForwardedRef.Wrapper.WasabyReactFn hasContainer="{{ _hasContainer }}"/>
    </ReactUnitTest._jsx.ForwardedRef.Wrapper.WasabyReactFn>
</ws:else>
