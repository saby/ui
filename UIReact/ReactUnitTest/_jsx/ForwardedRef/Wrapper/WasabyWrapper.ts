import { Control, TemplateFunction } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_jsx/ForwardedRef/Wrapper/WasabyWrapper';

export default class WasabyWrapper extends Control {
    protected _template: TemplateFunction = template;
    protected _hasContainer: boolean = false;

    protected _afterMount() {
        this._hasContainer = !!this._container;
    }
}
