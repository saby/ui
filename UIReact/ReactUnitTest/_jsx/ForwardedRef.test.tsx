/**
 * @jest-environment jsdom
 */
import { unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { Control, IControlOptions } from 'UI/Base';
import WasabyReactFn from './ForwardedRef/ReactFn/WasabyReactFn';
import WasabyReactClass from './ForwardedRef/ReactClass/WasabyReactClass';
import WasabyWrapper from './ForwardedRef/Wrapper/WasabyWrapper';

interface IOptions extends IControlOptions {
    contentOption?: boolean;
}

describe('ForwardedRef', () => {
    let container: HTMLDivElement;

    beforeEach(() => {
        container = document.createElement('div');
        document.body.appendChild(container);
        jest.useFakeTimers();
    });

    afterEach(() => {
        unmountComponentAtNode(container);
        container.remove();
        jest.useRealTimers();
        jest.restoreAllMocks();
    });

    // Wasaby контрол, который вставляет функциональный react компонент
    test('Wasaby > ReactFn', () => {
        act(() => {
            Control.createControl(WasabyReactFn, {}, container);
            jest.runAllTimers();
        });

        expect(container).toMatchSnapshot();
    });

    // Wasaby контрол, который вставляет функциональный react компонент,
    // в который в content вставляется другой функциональный react компонент
    test('Wasaby > ReactFn > content ReactFn', async () => {
        act(() => {
            Control.createControl<IOptions, HTMLDivElement>(
                WasabyReactFn,
                { contentOption: true },
                container
            );
            jest.runAllTimers();
        });

        expect(container).toMatchSnapshot();
    });

    // Wasaby контрол, который вставляет классовый react компонент
    test('Wasaby > ReactClass', () => {
        act(() => {
            Control.createControl(WasabyReactClass, {}, container);
            jest.runAllTimers();
        });

        expect(container).toMatchSnapshot();
    });

    // Wasaby контрол, который вставляет классовый react компонент,
    // в который в content вставляется другой классовый react компонент
    test('Wasaby > ReactClass > content ReactClass', () => {
        act(() => {
            Control.createControl<IOptions, HTMLDivElement>(
                WasabyReactClass,
                { contentOption: true },
                container
            );
            jest.runAllTimers();
        });

        expect(container).toMatchSnapshot();
    });

    test('Wasaby > Wasaby_React wrapper > Wasaby_React wrapper > Wasaby', () => {
        act(() => {
            Control.createControl<IOptions, HTMLDivElement>(
                WasabyWrapper,
                {},
                container
            );
            jest.runAllTimers();
        });

        expect(container).toMatchSnapshot();
    });

    test('Wasaby > Wasaby_React wrapper > Wasaby_React wrapper > content', () => {
        act(() => {
            Control.createControl<IOptions, HTMLDivElement>(
                WasabyWrapper,
                { contentOption: true },
                container
            );
            jest.runAllTimers();
        });

        expect(container).toMatchSnapshot();
    });
});
