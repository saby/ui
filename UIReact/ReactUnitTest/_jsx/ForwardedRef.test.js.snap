// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForwardedRef Wasaby > ReactClass > content ReactClass 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    ok
  </div>
</div>
`;

exports[`ForwardedRef Wasaby > ReactClass 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    ok
  </div>
</div>
`;

exports[`ForwardedRef Wasaby > ReactFn > content ReactFn 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    ok
  </div>
</div>
`;

exports[`ForwardedRef Wasaby > ReactFn 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    ok
  </div>
</div>
`;

exports[`ForwardedRef Wasaby > Wasaby_React wrapper > Wasaby_React wrapper > Wasaby 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div
    tabindex="0"
  >
    ok
  </div>
</div>
`;

exports[`ForwardedRef Wasaby > Wasaby_React wrapper > Wasaby_React wrapper > content 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div
    tabindex="0"
  >
    ok
  </div>
</div>
`;
